localstack/services/sns/filter.py

```diff
@@ -0,0 +1,429 @@
+        Although it's not documented anywhere, AWS allows `.` in the fields name in the filter policy and the payload,
+        and will evaluate them. However, it's not JSONPath compatible:
+        Example:
+        Policy: `{"field1.field2": "value1"}`
+        This policy will match both `{"field1.field2": "value1"}` and  {"field1: {"field2": "value1"}}`, unlike JSONPath
+        for which `.` points to a child node.
+        This might show they are flattening the both dictionaries to a single level for an easier matching without
+        recursion.
+        :param filter_policy: a dict, starting at the FilterPolicy
+        :param payload: a dict, starting at the MessageBody
+        :return: True if the payload respect the filter policy, otherwise False
+        """
+        flat_policy = self._flatten_dict(filter_policy)
+        flat_payloads = self._flatten_dict_with_list(payload)
+        for key, values in flat_policy.items():
+            if not any(
+                self._evaluate_condition(
+                    flat_payload.get(key), condition, field_exists=key in flat_payload
+                )
+                for condition in values
+                for flat_payload in flat_payloads
+            ):
+                return False
+        return True
+
+    def _evaluate_filter_policy_conditions_on_attribute(
+        self, conditions, attribute, field_exists: bool
+    ):
+        if not isinstance(conditions, list):
+            conditions = [conditions]
+
+        tpe = attribute.get("DataType") or attribute.get("Type") if attribute else None
+        val = attribute.get("StringValue") or attribute.get("Value") if attribute else None
+        if attribute is not None and tpe == "String.Array":
+            try:
+                values = json.loads(val)
+            except ValueError:
+                return False
+            for value in values:
+                for condition in conditions:
+                    if self._evaluate_condition(value, condition, field_exists):
+                        return True
+        else:
+            for condition in conditions:
+                value = val or None
+                if self._evaluate_condition(value, condition, field_exists):
+                    return True
+
+        return False
+
+    def _evaluate_condition(self, value, condition, field_exists: bool):
+        if not isinstance(condition, dict):
+            return field_exists and value == condition
+        elif (must_exist := condition.get("exists")) is not None:
+            # if must_exists is True then field_exists must be True
+            # if must_exists is False then fields_exists must be False
+            return must_exist == field_exists
+        elif value is None:
+            # the remaining conditions require the value to not be None
+            return False
+        elif anything_but := condition.get("anything-but"):
+            return value not in anything_but
```

Comment:
"anything-but" can also be a "prefix" https://docs.aws.amazon.com/sns/latest/dg/string-value-matching.html#string-anything-but-matching-prefix

Reply:
Yes! This is not supported yet, and we also do not support `suffix` and `equals-ignore-case`, which will be supported by #10691. I could try to sneak the `anything-but` into `prefix` in there as well.